src/io/flutter/perf/FlutterWidgetPerfManager.java

```diff
@@ -0,0 +1,295 @@
+      return;
+    }
+    trackRebuildWidgets = value;
+    onProfilingFlagsChanged();
+    if (debugIsActive && app != null && app.isSessionActive()) {
+      updateTrackWidgetRebuilds();
+    }
+  }
+
+  public boolean isTrackRepaintWidgets() {
+    return trackRepaintWidgets;
+  }
+
+  public void setTrackRepaintWidgets(boolean value) {
+    if (value == trackRepaintWidgets) {
+      return;
+    }
+    trackRepaintWidgets = value;
+    onProfilingFlagsChanged();
+    if (debugIsActive && app != null && app.isSessionActive()) {
+      updateTrackWidgetRepaints();
+    }
+  }
+
+  private void onProfilingFlagsChanged() {
+    if (currentStats != null) {
+      currentStats.setProfilingEnabled(isProfilingEnabled());
+    }
+  }
+
+  private boolean isProfilingEnabled() {
+    return trackRebuildWidgets || trackRepaintWidgets;
+  }
+
+  private void debugActive(Project project, FlutterViewMessages.FlutterDebugEvent event) {
+    debugIsActive = true;
+
+    assert(app != null);
+    app.addStateListener(this);
+    syncBooleanServiceExtension(TRACK_REBUILD_WIDGETS, () -> trackRebuildWidgets);
+    syncBooleanServiceExtension(TRACK_REPAINT_WIDGETS, () -> trackRepaintWidgets);
+
+    currentStats = new FlutterWidgetPerf(
+      isProfilingEnabled(),
+      new VmServiceWidgetPerfProvider(app),
+      (TextEditor textEditor) -> new EditorPerfDecorations(textEditor, app),
+      (TextEditor textEditor) -> new TextEditorFileLocationMapper(textEditor, app.getProject())
+    );
+  }
+
+  public void stateChanged(FlutterApp.State newState) {
+    switch (newState) {
+      case RELOADING:
+      case RESTARTING:
+        currentStats.clear();
+      break;
+      case STARTED:
+        notifyPerf();
+        break;
+    }
+  }
+
+  public void notifyAppRestarted() {
+    currentStats.clear();
+  }
+
+  public void notifyAppReloaded() {
+    currentStats.clear();
+  }
+
+  private void updateTrackWidgetRebuilds() {
+    app.maybeCallBooleanExtension(TRACK_REBUILD_WIDGETS, trackRebuildWidgets).whenCompleteAsync((v, e) -> {
+      notifyPerf();
+    });
+  }
+
+  private void updateTrackWidgetRepaints() {
+    app.maybeCallBooleanExtension(TRACK_REPAINT_WIDGETS, trackRepaintWidgets).whenCompleteAsync((v, e) -> {
+      notifyPerf();
+    });
+  }
+
+  private void syncBooleanServiceExtension(String serviceExtension, Computable<Boolean> valueProvider) {
+    streamSubscriptions.add(app.hasServiceExtension(serviceExtension, (supported) -> {
+      if (supported) {
+        app.callBooleanExtension(serviceExtension, valueProvider.compute());
+      }
+    }));
+  }
+
+  private boolean couldContainWidgets(@Nullable VirtualFile file) {
+    // TODO(jacobr): we might also want to filter for files not under the
+    // current project root.
+    return file != null && FlutterUtils.isDartFile(file);
+  }
+
+  private void updateCurrentAppChanged(@Nullable FlutterApp app) {
+    // TODO(jacobr): we currently only support showing stats for the last app
```

Comment:
I believe we do have a way to get what we consider to be the front-most app - based on the last active Debug view? `FlutterAppManager.getActiveApp()`, if that helps for this use case

Reply:
Expanded the comment. I think what we really want is to collect and aggregate thats from all running apps.